src/runtime/prte_finalize.c

```diff
@@ -74,7 +75,33 @@ int prte_finalize(void)
     /* release the cache */
     PRTE_RELEASE(prte_cache);
 
-    /* release the job hash table; pop the first element in key order and release it, repeat */
+    prte_display_prte_job_data("Finalize");
+
+    /* Release the job hash table
+     *
+     * First we need to iterate through an release all of the children.
```

Comment:
I think this works. You're clearing the list, which potentially loses any reference to a prte_job_t structure that hasn't been released, but since it's in a list that means it should he in the job data hash table. The second loop takes care of that, and now since all structures have been removed from the children lists, their reference count and belong-to fields have been cleared, so no more assert. If something does get missed, we're in finalize anyway.

Also, the wording in this comment reads a bit jumbled to me. Maybe something like:

There is the potential for a prte_job_data structure to still be in another prte_job_data structure's  children list, which will cause an assert when the first prte_job_data structure is released. Therefore, we traverse each prte_job_data's children list, removing any other prte_job_data structures from that list before iterating over hte job data hash table to release any remaining prte_job_data structures.

Reply:
Thanks. I'll clean up that comment.

Reply:
Done.